### PR TITLE
Don't use move semantics if not needed

### PR DESCRIPTION
**Pull request by jackcmay**
#### Problem

Lots of unnecessary vector allocations due to functions taking vectors instead of slices

#### Summary of Changes

Take a slice

Fixes #

### CODE REVIEW

File: sdk/src/message.rs [hunk 1]
```diff
@@ -198,7 +196,7 @@ impl Message {
         } = get_keys(&instructions, payer);
         let num_required_signatures = signed_keys.len() as u8;
         signed_keys.extend(&unsigned_keys);
-        let instructions = compile_instructions(instructions, &signed_keys);
+        let instructions = compile_instructions(&instructions, &signed_keys);
```
Comment (CriesofCarrots) [inline sdk/src/message.rs]: Added `&` in this line is unneeded (also can remove `&` from the call to `get_keys()` immediately above)

Comment (jackcmay) [reply sdk/src/message.rs]: Thanks, I must have made those changes before I ended up changing `new_with_payer`, suprised that clippy didn't complain.

Comment (CriesofCarrots) [reply sdk/src/message.rs]: Yeah, it is surprisingly permissive of double references in some contexts, and we have a quite a few in our code base. Discovered that in one of the more heavy-handed Signer trait-object iterations.

**Comment by jackcmay** [comment]
@mvines  Need this in v1.0?

**Comment by jackcmay** [comment]
I'm sure, not a game changer but helps

**Comment by mvines** [comment]
> @mvines Need this in v1.0?

I don't see a strict need, but bringing this into v1.0 should help avoid future merge conflicts given how sweeping the changes are so I'm for bringing it in to hopefully reduce our support burden until v1.1.x moves to the stable release channel (end of April)

**Pull request merged**